core/src/main/scala/org/apache/spark/SparkEnv.scala

```diff
@@ -368,10 +374,12 @@ object SparkEnv extends Logging {
           } else {
             None
           }, blockManagerInfo,
-          mapOutputTracker.asInstanceOf[MapOutputTrackerMaster])),
+          mapOutputTracker.asInstanceOf[MapOutputTrackerMaster],
+          recentlyRemovedExecutors)),
       registerOrLookupEndpoint(
         BlockManagerMaster.DRIVER_HEARTBEAT_ENDPOINT_NAME,
-        new BlockManagerMasterHeartbeatEndpoint(rpcEnv, isLocal, blockManagerInfo)),
+        new BlockManagerMasterHeartbeatEndpoint(rpcEnv, isLocal, blockManagerInfo,
+          recentlyRemovedExecutors)),
       conf,
```

Comment:
Move the cache into `BlockManagerMasterEndpoint`.
We should fix the interface here ... `BlockManagerMasterHeartbeatEndpoint` simply needs a way to validate if an executor was recently removed - does not need to know if it was a Cache/Set/etc.

Reply:
Sure.
The same comment applies to already present `blockManagerInfo` map as well.
I could refactor that as well. I was thinking of creating a `BlockMangerEndpointSharedState` class which contains `blockManagerInfo` and `recentlyRemovedExecutors`. The class would expose corresponding methods for lookup and updates.